### PR TITLE
Ignored comments in interpreter before reordering

### PR DESCRIPTION
**Pull request by Heulitig**
- reverted the reorder() in utils.rs, it doesn't ignore commented sections anymore. 
- commented things are ignored in a separate function before reorder() is called
- after calling this function, the ParsedDocument::sections will only have uncommented portion of the original parsed document.

### CODE REVIEW

File: src/p2/interpreter.rs [hunk 1]
```diff
@@ -741,6 +745,102 @@ impl ParsedDocument {
         self.processing_imports = false;
     }
 
+    fn ignore_comments(&mut self) -> ftd::p1::Result<()> {
+        let mut processed_sections: Vec<ftd::p1::Section> = Vec::new();
+        for section in self.sections.iter() {
```
Comment (Arpita-Jaiswal) [inline src/p2/interpreter.rs:750]:
```suggestion
        for section in self.sections.iter_mut() {
```

File: src/p2/interpreter.rs [hunk 2]
```diff
@@ -741,6 +745,102 @@ impl ParsedDocument {
...
+
+            let mut filtered_section = ftd::p1::Section {
+                name: section.name.clone(),
+                caption: section.caption.clone(),
+                header: Default::default(),
+                body: None,
+                sub_sections: Default::default(),
+                is_commented: section.is_commented,
+                line_number: section.line_number,
+            };
```
Comment (Arpita-Jaiswal) [inline src/p2/interpreter.rs:764]: Use section as `mut` only. `.iter_mut()` would do it. No need to create this variable

**Review by Arpita-Jaiswal — CHANGES_REQUESTED** [review]

**Comment by Arpita-Jaiswal** [comment]
@Heulitig Search for all the places from where the comments are ignored and remove them. Also that places where escape comment is handled, remove them too.

Make sure everything is working fine with your changes

File: src/p2/interpreter.rs [hunk 3]
```diff
@@ -70,9 +70,15 @@ impl InterpreterState {
             }
             self.document_stack[l].done_processing_imports();
 
+            // println!("\nBefore ignoring comments !! ===========================================\n");
+            // self.document_stack[l].show_document();
+
```
Comment (amitu) [inline src/p2/interpreter.rs:75]: Don't add these comments, delete them.

File: src/p2/interpreter.rs [hunk 4]
```diff
@@ -745,24 +752,44 @@ impl ParsedDocument {
         self.processing_imports = false;
     }
 
+    /// prints the parsed document (for debugging purposes)
+    fn show_document(&self) {
+        for section in self.sections.iter() {
+            println!(
+                "section name: {}, caption: {:?}, is_commented: {}",
+                &section.name, &section.caption, section.is_commented
+            );
```
Comment (amitu) [inline src/p2/interpreter.rs:762]: Use `dbg!()` for debugging, not print. It is less typing and gives more info, and it makes it obvious to everyone that it is a debug statement and not an actual requirement.

Comment (Heulitig) [reply src/p2/interpreter.rs:762]: Ohh okay

File: src/p2/interpreter.rs (hunk 3 above)
Comment (Heulitig) [reply src/p2/interpreter.rs:75]: I was just testing this I will remove these comments and that debug function once all conditions are passed in the final commit.

File: src/p1/section.rs [hunk 5]
```diff
@@ -36,18 +36,7 @@ impl Section {
...
-                _ if body.1.starts_with(r"\/") =>
-                {
-                    #[allow(clippy::single_char_pattern)]
-                    Some((body.0, body.1.strip_prefix(r"\").expect("").to_string()))
-                }
-                _ if body.1.starts_with('/') => None,
-                _ => self.body.clone(),
-            },
-        };
+        let body = self.body_without_comment();
```
Comment (Arpita-Jaiswal) [inline src/p1/section.rs:39]: Why is this needed? Is this not already taken care of in `ignore_comments`? If not so, please do it then. This should be just the `body`. if there's any redundant code, please remove them.

File: src/p1/sub_section.rs [hunk 6]
```diff
@@ -59,18 +59,7 @@ impl SubSection {
...
-                _ if body.1.starts_with(r"\/") =>
-                {
-                    #[allow(clippy::single_char_pattern)]
-                    Some((body.0, body.1.strip_prefix(r"\").expect("").to_string()))
-                }
-                _ if body.1.starts_with('/') => None,
-                _ => self.body.clone(),
-            },
-        };
+        let body = self.body_without_comment();
```
Comment (Arpita-Jaiswal) [inline src/p1/sub_section.rs:62]: Same as above

File: src/p2/interpreter.rs [hunk 7]
```diff
@@ -45,6 +45,11 @@ impl InterpreterState {
 
         let l = self.document_stack.len() - 1; // Get the top of the stack
 
+        // Any of these two functions can be used to remove comments from parsed document
+        // self.document_stack[l].ignore_all_comments()?;
```
Comment (Arpita-Jaiswal) [inline src/p2/interpreter.rs:49]: Choose the best and keep only one

File: src/p2/interpreter.rs [hunk 8]
```diff
@@ -741,6 +737,157 @@ impl ParsedDocument {
...
+            // Filter section body
+            section.body = {
+                match section.body {
+                    Some(ref b) if b.1.trim().is_empty() => None,
+                    Some(ref b) if b.1.trim().starts_with('/') => None,
+                    // To allow '/content' in section body, we need to use "\/content"
+                    // while stripping out the initial '\' from this body
+                    Some(ref b) if b.1.trim().starts_with(r"\/") => {
+                        Some((b.0, b.1.trim().replacen(r"\", "", 1)))
+                    }
```
Comment (Arpita-Jaiswal) [inline src/p2/interpreter.rs:832]: Check this for `headers` and `caption` too. Also use one function which does this in all headers, body and caption in sections and subsections.

**Review by Arpita-Jaiswal — CHANGES_REQUESTED** [review]

File: src/p2/interpreter.rs [hunk 9]
```diff
@@ -741,6 +737,157 @@ impl ParsedDocument {
...
+                );
+
+                for (_ln, key, value) in subsection.header.0.iter() {
+                    dbg!(key, value);
+                }
+
+                dbg!(&subsection.body);
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Heulitig) [reply src/p2/interpreter.rs]: removed

File: src/p2/interpreter.rs [hunk 10]
```diff
@@ -741,6 +737,157 @@ impl ParsedDocument {
...
+    #[allow(dead_code)]
+    fn ignore_all_comments(&mut self) -> ftd::p1::Result<()> {
+        self.sections = self
+            .sections
+            .iter()
+            .filter(|s| !s.is_commented)
+            .map(|s| s.remove_comments())
+            .collect::<Vec<ftd::p1::Section>>();
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (Heulitig) [reply src/p2/interpreter.rs]: This is used now, the other one is removed

File: src/p1/section.rs (hunk 5 above)
Comment (Heulitig) [reply src/p1/section.rs]: body_without_comment() is only used by remove_comments() which in turn is only used by ignore_comments() in the interpreter.

File: src/p2/interpreter.rs (hunk 7 above)
Comment (Heulitig) [reply src/p2/interpreter.rs]: I have removed the other one

File: src/p1/sub_section.rs (hunk 6 above)
Comment (Heulitig) [reply src/p1/sub_section.rs]: Same as above

File: src/p2/interpreter.rs [hunk 11]
```diff
@@ -741,6 +737,157 @@ impl ParsedDocument {
...
+    ///
+    /// 5.  /sub-section-header: value
+    ///
+    /// ## Note: To allow ["/content"] inside body, use ["\\/content"].
+    ///
+    /// Only '/' comments are ignored here.
+    /// ';' comments are ignored inside the [`parser`] itself.
+    ///
+    /// [`parser`]: ftd::p1::parser::parse
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (Heulitig) [reply src/p2/interpreter.rs]: This just creates a link to that function in the functional doc within the crate in case if anyone wants to refer its definition.

File: src/p2/interpreter.rs (hunk 8 above)
Comment (Heulitig) [reply src/p2/interpreter.rs]: Done
